src/Common/src/Interop/OSX/Interop.libproc.cs

```diff
@@ -0,0 +1,542 @@
+        /// to not having enough permissions to query for the data of that specific process
+        /// </returns>
+        [DllImport(Interop.Libraries.libproc, SetLastError = true)]
+        private static unsafe extern int proc_pidinfo(
+            int pid,
+            int flavor,
+            ulong arg,
+            ulong* buffer,
+            int bufferSize);
+
+        /// <summary>
+        /// Gets the process information for a given process
+        /// </summary>
+        /// <param name="pid">The PID (process ID) of the process</param>
+        /// <returns>
+        /// Returns a valid proc_taskallinfo struct for valid processes that the caller 
+        /// has permission to access; otherwise, returns null
+        /// </returns>
+        internal static unsafe proc_taskallinfo? GetProcessInfoById(int pid)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            // Get the process information for the specified pid
+            int size = Marshal.SizeOf<proc_taskallinfo>();
+            proc_taskallinfo info = default(proc_taskallinfo);
+            int result = proc_pidinfo(pid, PROC_PIDTASKALLINFO, 0, &info, size);
+            return (result == size ? new proc_taskallinfo?(info) : null);
+        }
+
+        /// <summary>
+        /// Gets the thread information for the given thread
+        /// </summary>
+        /// <param name="thread">The ID of the thread to query for information</param>
+        /// <returns>
+        /// Returns a valid proc_threadinfo struct for valid threads that the caller
+        /// has permissions to access; otherwise, returns null
+        /// </returns>
+        internal static unsafe proc_threadinfo? GetThreadInfoById(int pid, ulong thread)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            // Negative TIDs are invalid
+            if (thread < 0)
+            {
+                throw new ArgumentOutOfRangeException("thread");
+            }
+
+            // Get the thread information for the specified thread in the specified process
+            int size = Marshal.SizeOf<proc_threadinfo>();
+            proc_threadinfo info = default(proc_threadinfo);
+            int result = proc_pidinfo(pid, PROC_PIDTHREADINFO, (ulong)thread, &info, size);
+            return (result == size ? new proc_threadinfo?(info) : null);
+        }
+
+        internal static unsafe List<KeyValuePair<ulong, proc_threadinfo?>> GetAllThreadsInProcess(int pid)
+        {
+            // Negative PIDs are invalid
+            if (pid < 0)
+            {
+                throw new ArgumentOutOfRangeException("pid");
+            }
+
+            int result = 0;
+            int size = 20; // start assuming 20 threads is enough
```

Comment:
Yeah, I was mostly interested in the average, i.e. how many threads to most processes use on average.  it's not a big deal, but if it's easy for you to write a quick test that iterates through all of the processes and prints out some basic stats, like min, average, and max, that's be instructive.  Don't block on it though; if there's an optimization to be made, we can always do so later.